### PR TITLE
Запрос рекомендации

### PR DESCRIPTION
**Pull request by Weltschmerz99**

### CODE REVIEW

**Review by Vahe650 — APPROVED** [review]

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import school.faang.user_service.service.RecommendationRequestService;
+
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.List;
+import java.time.LocalDateTime;
+import java.util.stream.Collectors;
+
+@Controller
+@Validated
```
Comment (Vahe650) [inline src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: @Validated remove please

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+public class RecommendationRequestController {
+    private final RecommendationRequestService recommendationRequestService;
+
+    @Autowired
+    public RecommendationRequestController(RecommendationRequestService recommendationRequestService){
+        this.recommendationRequestService = recommendationRequestService;
+    }
+
+    // Метод для отправки запроса на рекомендацию
+    public RecommendationRequestDto requestRecommendation(RecommendationRequestDto recommendationRequestDto) {
```
Comment (Vahe650) [inline src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: public RecommendationRequestDto requestRecommendation(@Valid RecommendationRequestDto recommendationRequestDto) {

**Review by Vahe650 — APPROVED** [review]

File: src/main/java/school/faang/user_service/dto/RejectionDto.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package school.faang.user_service.dto;
+
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/dto/RejectionDto.java:2]: you can use Lombok annotations

File: src/main/java/school/faang/user_service/service/RecommendationRequestService.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import school.faang.user_service.repository.recommendation.RecommendationRequestRepository;
+import school.faang.user_service.repository.recommendation.SkillRequestRepository;
+
+@Service
+public class RecommendationRequestService {
+    private final RecommendationRequestRepository recommendationRequestRepository;
+    private final SkillRequestRepository skillRequestRepository;
+    private final StringToLongMapper stringToLongMapper;
+
+    @Autowired
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/service/RecommendationRequestService.java:26]: Lombok ,as well, would be helpful

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            throw new IllegalArgumentException("Сообщение не может быть пустым");
+        }
+
+        // Извлечение данных из DTO запроса
+        Long requesterId = recommendationRequestDto.getRequesterId();
+        Long receiverId = recommendationRequestDto.getReceiverId();
+        List<String> skills = recommendationRequestDto.getSkills();
+
+        // Создание нового DTO и вызов метода создания запроса в сервисе
+        RecommendationRequestDto dto = new RecommendationRequestDto();
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:38]: why do you create new Dto and never use it?

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        recommendationRequestService.create(requesterId, receiverId, dto);
+
+        return recommendationRequestDto;
+    }
+
+    // Метод для получения списка запросов на рекомендацию
+    public List<RecommendationRequestDto> getRecommendationRequests(RequestFilterDto filter) {
+        // Получение списка запросов и их преобразование в DTO
+        return recommendationRequestService.getRequests(filter)
+                .stream()
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: you can create additional method in Mapper to map List<Entity> into List<Dto>

File: src/main/java/school/faang/user_service/mapper/StringToLongMapper.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.factory.Mappers;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+@Mapper
+@Component
+public interface StringToLongMapper {
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/mapper/StringToLongMapper.java:14]: It is better to watch video on platform, how to do that properly

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                .map(this::mapToDto)
+                .collect(Collectors.toList());
+    }
+
+    // Приватный вспомогательный метод для преобразования сущности запроса в DTO
+    private RecommendationRequestDto mapToDto(RecommendationRequest recommendationRequest) {
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: it is better to that in Mapper class

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+        return dto;
+    }
+
+    // Метод для получения конкретного запроса на рекомендацию
+    @GetMapping("/{id}")
+    public ResponseEntity<RecommendationRequestDto> getRecommendationRequest(@PathVariable long id) {
+        // Получение запроса по идентификатору и преобразование его в DTO
+        RecommendationRequest recommendationRequest = recommendationRequestService.getRequest(id);
+        if (recommendationRequest == null) {
```
Comment (gutorov) [inline src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: you have already checked if the recommendation is null in service, you do not need to do that twice

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java (hunk 2 above)
Comment (Weltschmerz99) [reply src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: fixed

File: src/main/java/school/faang/user_service/dto/RejectionDto.java (hunk 3 above)
Comment (Weltschmerz99) [reply src/main/java/school/faang/user_service/dto/RejectionDto.java:2]: Lombok doesn't always work, so I use a safer option

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java (hunk 5 above)
Comment (Weltschmerz99) [reply src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:38]: i use it in the next line

File: src/main/java/school/faang/user_service/mapper/StringToLongMapper.java (hunk 7 above)
Comment (Weltschmerz99) [reply src/main/java/school/faang/user_service/mapper/StringToLongMapper.java:14]: updated

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java (hunk 8 above)
Comment (Weltschmerz99) [reply src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: fixed

File: src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java (hunk 9 above)
Comment (Weltschmerz99) [reply src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java]: fixed

**Review by Vahe650 — APPROVED** [review]

**Pull request merged**